plugin/servergen.go

```diff
@@ -45,61 +44,217 @@ func (p *OrmPlugin) generateDefaultServer(file *generator.FileDescriptor) {
+			if _, exists := convertibleTypes[outTypeName]; exists {
+				typeOrmable = true
+			}
+		}
+	}
+	if hasID && typeOrmable {
+		return true, lintName(outTypeName)
+	}
+	return false, ""
 }
 
 func (p *OrmPlugin) generateUpdateServerMethod(service *descriptor.ServiceDescriptorProto, method *descriptor.MethodDescriptorProto) {
 	inType, outType, methodName, svcName := p.getMethodProps(service, method)
 	p.generateMethodSignature(inType, outType, methodName, svcName)
-	p.P(`var out `, p.TypeName(outType))
-	p.P(`res, err := DefaultUpdate`, strings.TrimPrefix(methodName, "Update"), `(ctx, in.GetPayload(), m.DB)`)
-	p.P(`out.Result = res`)
-	p.P(`return &out, err`)
-	p.P(`}`)
+	follows, typeName := p.followsUpdateConventions(inType, outType)
+	if follows {
+		p.P(`res, err := DefaultUpdate`, typeName, `(ctx, in.GetPayload(), m.DB)`)
+		p.P(`if err != nil {`)
+		p.P(`return nil, err`)
+		p.P(`}`)
+		p.P(`return &`, p.TypeName(outType), `{Result: res}, nil`)
+		p.P(`}`)
+	} else {
+		p.generateEmptyBody(outType)
+	}
+}
+
+func (p *OrmPlugin) followsUpdateConventions(inType generator.Object, outType generator.Object) (bool, string) {
+	inMsg := inType.(*generator.Descriptor)
+	outMsg := outType.(*generator.Descriptor)
+	var inTypeName string
+	var typeOrmable bool
+	for _, field := range inMsg.Field {
+		if field.GetName() == "payload" {
+			gType, _ := p.GoType(inMsg, field)
+			inTypeName = strings.TrimPrefix(gType, "*")
+			if _, exists := convertibleTypes[inTypeName]; exists {
+				typeOrmable = true
+			}
+		}
+	}
+	var outTypeName string
+	for _, field := range outMsg.Field {
+		if field.GetName() == "result" {
+			gType, _ := p.GoType(outMsg, field)
+			outTypeName = strings.TrimPrefix(gType, "*")
+		}
+	}
+	if inTypeName == outTypeName && typeOrmable {
+		return true, lintName(inTypeName)
+	}
+	return false, ""
 }
 
 func (p *OrmPlugin) generateDeleteServerMethod(service *descriptor.ServiceDescriptorProto, method *descriptor.MethodDescriptorProto) {
 	inType, outType, methodName, svcName := p.getMethodProps(service, method)
 	p.generateMethodSignature(inType, outType, methodName, svcName)
-	p.P(`return nil, DefaultDelete`, strings.TrimPrefix(methodName, "Delete"), `(ctx, in.GetPayload(), m.DB)`)
-	p.P(`}`)
+	follows, typeName := p.followsDeleteConventions(inType, outType, method)
+	if follows {
+		p.P(`return &`, p.TypeName(outType), `{}, `, `DefaultDelete`, typeName, `(ctx, &`, typeName, `{Id: in.GetId()}, m.DB)`)
+		p.P(`}`)
+	} else {
+		p.generateEmptyBody(outType)
+	}
+}
+
+func (p *OrmPlugin) followsDeleteConventions(inType generator.Object, outType generator.Object, method *descriptor.MethodDescriptorProto) (bool, string) {
+	inMsg := inType.(*generator.Descriptor)
+	var hasID bool
+	for _, field := range inMsg.Field {
+		if field.GetName() == "id" {
+			hasID = true
+		}
+	}
+	var typeName string
+	v, err := proto.GetExtension(method.GetOptions(), gorm.E_Method)
+	opts := v.(*gorm.MethodOptions)
```

Comment:
This line panics if (gorm.server).autogen is set, but (gorm.method).object_type is not set. Use the opts, ok := v.(*gorm.MethodOptions) to avoid this. If it needs to fail in that scenario, something like p.Fail("Cannot generate Delete handler if object_type option is not set in rpc definition") would be good.